[MLIR] Generalize expand_shape to take shape as explicit input

Shukla-Gaurav opened the pull request:
This patch generalizes tensor.expand_shape and memref.expand_shape to consume the output shape as a list of SSA values.  This enables us to implement generic reshape operations with dynamic shapes using collapse_shape/expand_shape pairs.

The output_shape input to expand_shape follows the static/dynamic representation that's also used in `tensor.extract_slice`.

Differential Revision: https://reviews.llvm.org/D140821

ramiro050 committed:
[MLIR] Generalize expand_shape to take shape as explicit input

This patch generalizes tensor.expand_shape and memref.expand_shape to consume
the output shape as a list of SSA values.  This enables us to implement generic
reshape operations with dynamic shapes using collapse_shape/expand_shape pairs.

The output_shape input to expand_shape follows the static/dynamic representation
that's also used in `tensor.extract_slice`.

Differential Revision: https://reviews.llvm.org/D140821

Shukla-Gaurav commented:
@MaheshRavishankar @joker-eph This PR is an exact copy of https://github.com/llvm/llvm-project/pull/69267 with one line change(`mlir/lib/Dialect/Utils/CMakeLists.txt`) to fix the bot failure. Could you please approve this?

Shukla-Gaurav commented:
@bjacob fyi

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:
```diff
@@ -750,6 +750,31 @@ static void updateExpandedGenericOpRegion(PatternRewriter &rewriter,
+      continue;
+    bool foundDynamic = false;
+    for (int64_t shape : expandedShape) {
+      if (ShapedType::isDynamic(shape)) {
+        if (foundDynamic) {
```
Nit: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

ftynse commented on mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp:
```diff
@@ -750,6 +750,31 @@ static void updateExpandedGenericOpRegion(PatternRewriter &rewriter,
+    for (int64_t shape : expandedShape) {
+      if (ShapedType::isDynamic(shape)) {
+        if (foundDynamic) {
+          return rewriter.notifyMatchFailure(
+              linalgOp, "cannot infer expanded shape with multiple dynamic "
+                        "dims in the same reassociation group");
```
We usually have tests for diagnostic messages. Is there a way to test this?

Shukla-Gaurav replied:
I am not sure how to add a test for this diagnostic message. This utility function puts an early check to detect cases where fusion by expansion is not possible.

ftynse commented on mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:
```diff
@@ -2280,14 +2332,28 @@ LogicalResult ExpandShapeOp::verify() {
     return emitOpError("expected expanded type to be ")
            << *expectedResultType << " but found " << resultType;
 
+  if ((int64_t)getStaticOutputShape().size() != resultType.getRank())
+    return emitOpError("expected number of static shape bounds to be equal to "
+                       "the output rank (")
+           << resultType.getRank() << ") but found "
+           << getStaticOutputShape().size() << " inputs instead";
+
+  if ((int64_t)getOutputShape().size() !=
+      llvm::count(getStaticOutputShape(), ShapedType::kDynamic))
+    return emitOpError("mismatch in dynamic dims in output_shape and "
+                       "static_output_shape: static_output_shape has ")
+           << llvm::count(getStaticOutputShape(), ShapedType::kDynamic)
+           << " dynamic dims while output_shape has " << getOutputShape().size()
+           << " values";
+
```
We should have tests for these.

ftynse replied:
You mean the assembly syntax? Is it still possible to trigger this message by writing the op using the generic syntax? If not, it cannot be triggered on user-supplied IR and should become an assertion.

ftynse commented on mlir/lib/Dialect/Utils/CMakeLists.txt:
```diff
@@ -8,5 +8,6 @@ add_mlir_library(MLIRDialectUtils
   MLIRDialectUtilsIncGen
 
   LINK_LIBS PUBLIC
+  MLIRArithDialect
```
Dialect/Utils must not depend on a specific dialect.

ftynse commented on mlir/lib/Dialect/Utils/ReshapeOpsUtils.cpp:
```diff
@@ -8,6 +8,7 @@
 
 #include "mlir/Dialect/Utils/ReshapeOpsUtils.h"
 
+#include "mlir/Dialect/Arith/IR/Arith.h"
```
Dialect/Utils must not depend on a specific dialect.

MaheshRavishankar replied:
Makes sense. The only reasonable place to put this would be in arith dialect utils.
Kind of a strange place for reshape utils, but the way it works out, it just needs to go there

Shukla-Gaurav replied:
Moved to Arith/Utils. Thanks :)

ftynse reviewed (changes requested):
I have concerns with layering here. This adds a dependency on the arithmetic dialect from dialect utils. It's not exactly clear where such a code should go since it seems to be shared between memref and tensor, maybe somewhere next to type inference.

Incidentally, this dependency is also what broke the build when this was first landed.

MaheshRavishankar commented:
@ftynse unless this arith -> utils dependency breaks builds (I dont think it does, cause there is no cyclic dependency afaics), i think it might be better to land this and then remove the dependency as a quick follow up. This is fairly big change, and more things might get added that will have to be resolved. I am inclined to take the snap shot as is, but fix the dependency issue quickly after.

Commit:
[MLIR] Add test cases for illegal output shape argument

Signed-Off-by: Gaurav Shukla<gaurav.shukla@amd.com>

Commit:
[MLIR] Remove ArithDialect dependency from Dialect/Utils

This commit moves the inferExpandShapeOutputShape utility from the
Dialect/Utils/ReshapeOpsUtils.cpp to Arith/Utils/Utils.cpp in order to
remove specific dialect dependencies from the DialectUtils.

Signed-Off-by: Gaurav Shukla <gaurav.shukla@amd.com>

ftynse commented:
> @ftynse unless this arith -> utils dependency breaks builds (I dont think it does, cause there is no cyclic dependency afaics), i think it might be better to land this and then remove the dependency as a quick follow up. 

I don't see a cyclic dependency immediately. This is a bit of a bikeshedding discussion, one can simply move a file under a new `ShapedTypeArithUtils` and resolve the surface concern: `DialectUtils` suggests that it's something cheap and useful for all dialects, but it comes with a dependency on Arith (and UB, etc.). There's a deeper concern: we don't have a dialect-agnostic way to create constants, so anything that creates constants will have to depend on Arith or similar. I certainly won't block this on resolving that concern.

> This is fairly big change, and more things might get added that will have to be resolved. I am inclined to take the snap shot as is, but fix the dependency issue quickly after.

If the issue can be fixed quickly, why not quickly add the solution to the PR? :)

My worry here is by allowing this dependency temporarily, more uses will creep in, making it harder to remove as the time goes. I have made a similar comment this week on another PR...

How about we ask a third opinion, @joker-eph ?

Shukla-Gaurav commented:
> > @ftynse unless this arith -> utils dependency breaks builds (I dont think it does, cause there is no cyclic dependency afaics), i think it might be better to land this and then remove the dependency as a quick follow up.
> 
> I don't see a cyclic dependency immediately. This is a bit of a bikeshedding discussion, one can simply move a file under a new `ShapedTypeArithUtils` and resolve the surface concern: `DialectUtils` suggests that it's something cheap and useful for all dialects, but it comes with a dependency on Arith (and UB, etc.). There's a deeper concern: we don't have a dialect-agnostic way to create constants, so anything that creates constants will have to depend on Arith or similar. I certainly won't block this on resolving that concern.
> 
> > This is fairly big change, and more things might get added that will have to be resolved. I am inclined to take the snap shot as is, but fix the dependency issue quickly after.
> 
> If the issue can be fixed quickly, why not quickly add the solution to the PR? :)
> 
> My worry here is by allowing this dependency temporarily, more uses will creep in, making it harder to remove as the time goes. I have made a similar comment this week on another PR...
> 
> How about we ask a third opinion, @joker-eph ?

The utility has been moved to Arith/Utils, so no specific dialect dependency on the utils dialect. Let me know if Arith/Utils is the right place to put the utility `inferExpandShapeOutputShape`?

ftynse commented:
> The utility has been moved to Arith/Utils, so no specific dialect dependency on the utils dialect. Let me know if Arith/Utils is the right place to put the utility inferExpandShapeOutputShape?

Maybe not, but it's a separate discussion. Thanks for addressing my concerns!

joker-eph commented:
Thanks @ftynse

joker-eph commented:
This broke the bot here: https://lab.llvm.org/buildbot/#/builders/264/builds/10013

Unless you have a fix available, please revert.

PeimingLiu commented:
FYI, tests related to sparse tensor dialect should be fixed by https://github.com/llvm/llvm-project/pull/90637

joker-eph commented:
Priority is always getting the bot back green: so either merge immediately or revert the original PR in the meantime.

PeimingLiu commented:
Merged, but there is one more remaining outside sparse tensor tests.

hanhanW commented:
> Merged, but there is one more remaining outside sparse tensor tests.

Testing now, https://github.com/llvm/llvm-project/pull/90649 should fix the issue.

hanhanW commented:
side question: are we able to enable integration tests in presubmit? All the integration tests are marked `UNSUPPORTED` in https://buildkite.com/llvm-project/github-pull-requests/builds/59998#018f2d98-8ac6-47ac-8d79-65d3e7cd6c28

hanhanW commented:
> > Merged, but there is one more remaining outside sparse tensor tests.
> 
> Testing now, #90649 should fix the issue.

I verified that it is passing with the PR.
